### PR TITLE
Idata deduplication

### PR DESCRIPTION
**Pull request by ravinderjangra**

### CODE REVIEW

File: src/data_handler.rs [hunk 1]
```diff
@@ -236,6 +242,46 @@ impl DataHandler {
             }
         }
     }
+
+    pub fn handle_node_left_action(&mut self, node_left: XorName) -> Option<Vec<Action>> {
```
Comment (lionel1704) [inline src/data_handler.rs:246]: Needs rename.

File: src/data_handler/idata_handler.rs [hunk 2]
```diff
@@ -35,6 +35,7 @@ struct ChunkMetadata {
 
 pub(super) struct IDataHandler {
     id: NodePublicId,
+    idata_copy_ops: BTreeSet<MessageId>,
     idata_ops: BTreeMap<MessageId, IDataOp>,
```
Comment (lionel1704) [inline src/data_handler/idata_handler.rs:39]: We can differentiate these as, `idata_client_ops` and `idata_elder_ops`

File: src/data_handler/idata_handler.rs [hunk 3]
```diff
@@ -209,6 +213,38 @@ impl IDataHandler {
         }
     }
 
+    pub(super) fn get_idata_copy(
```
Comment (lionel1704) [inline src/data_handler/idata_handler.rs:216]: Rename.

File: src/data_handler/idata_handler.rs [hunk 4]
```diff
@@ -266,6 +303,17 @@ impl IDataHandler {
         }
     }
 
+    pub fn check_idata_holders(
```
Comment (lionel1704) [inline src/data_handler/idata_handler.rs:306]: Rename.

File: src/data_handler/idata_handler.rs [hunk 5]
```diff
@@ -336,16 +390,32 @@ impl IDataHandler {
...
+            info!("DataHandler: Forward PutIdata copy request");
+            let _ = self.idata_copy_ops.remove(&message_id);
+            self.remove_idata_op_if_concluded(&message_id)
+                .map(|idata_op| Action::RespondToOurDataHandlers {
+                    target: sender,
+                    rpc: Rpc::Response {
+                        requester: idata_op.client().clone(),
+                        response: Response::Mutation(Ok(())),
+                        message_id,
+                        refund: None,
```
Comment (lionel1704) [inline src/data_handler/idata_handler.rs:403]: Who are we responding to here?

File: src/data_handler/idata_handler.rs [hunk 6]
```diff
@@ -336,16 +390,32 @@ impl IDataHandler {
...
+                    },
+                })
+        } else {
+            info!("ClientHandler: Forward PutIdata request");
```
Comment (lionel1704) [inline src/data_handler/idata_handler.rs:407]: This log message seems wrong. The PutIData flow is complete in this step and we respond to the client's elders with `Ok(())`

File: src/data_handler/idata_handler.rs [hunk 7]
```diff
@@ -425,13 +495,107 @@ impl IDataHandler {
...
+                                    request: Request::IData(idata_op.request().clone()),
+                                    requester: us_as_requester,
+                                    message_id: new_msg_id,
+                                },
+                            })
+                        }
+                    }
+                }
+                Err(_) => None,
+            }
```
Comment (lionel1704) [inline src/data_handler/idata_handler.rs:538]: Why do we need this here?

File: src/data_handler/idata_handler.rs [hunk 8]
```diff
@@ -425,13 +495,107 @@ impl IDataHandler {
...
+            self.idata_op_mut(&message_id).and_then(|idata_op| {
+                idata_op.handle_get_idata_resp(sender, result, &own_id, message_id)
+            })
+        };
+
         let _ = self.remove_idata_op_if_concluded(&message_id);
         action
     }
 
+    fn get_metadata_for_all_chunks(
```
Comment (lionel1704) [inline src/data_handler/idata_handler.rs:549]: Rename: `get_metadata_for_chunks_held_by(holder_name)` or something similar.

File: src/data_handler/idata_handler.rs [hunk 9]
```diff
@@ -425,13 +495,107 @@ impl IDataHandler {
...
+        &mut self,
+        holder: XorName,
+    ) -> NdResult<BTreeMap<IDataAddress, BTreeSet<XorName>>> {
+        let mut idata_addresses: BTreeMap<IDataAddress, BTreeSet<XorName>> = BTreeMap::new();
+        // Get all idata addresses and holders when any holder left the network
+        for kv in self.metadata.iter() {
+            match kv.get_value::<ChunkMetadata>() {
+                None => {
+                    warn!("{}: is not responsible for this chunk", holder);
+                }
```
Comment (lionel1704) [inline src/data_handler/idata_handler.rs:559]: When is chunk metadata `None`. We will need to change this log message accordingly.

File: src/data_handler/idata_handler.rs [hunk 10]
```diff
@@ -482,8 +646,7 @@ impl IDataHandler {
         None
     }
 
-    // Returns an iterator over all of our section's non-full adults' names, sorted by closest to
-    // `target`.
+    // Returns XorName's for the target holders for any idata chunck.
```
Comment (lionel1704) [inline src/data_handler/idata_handler.rs:649]:
```suggestion
    // Returns `XorName`s of the target holders for an idata chunk.

```

File: src/data_handler/idata_handler.rs [hunk 11]
```diff
@@ -482,8 +646,7 @@ impl IDataHandler {
...
     fn make_holder_list_for_idata(&self, target: &XorName) -> Vec<XorName> {
```
Comment (lionel1704) [inline src/data_handler/idata_handler.rs:650]: Rename.

File: src/data_handler/idata_handler.rs [hunk 12]
```diff
@@ -506,6 +674,30 @@ impl IDataHandler {
             closest_adults
         }
     }
+
+    // Returns XorName's for the target holders for any idata chunck copy.
+    fn make_holder_list_for_idata_copy(&self, target: &IDataAddress) -> BTreeSet<XorName> {
```
Comment (lionel1704) [inline src/data_handler/idata_handler.rs:679]: Rename.

File: src/data_handler/idata_handler.rs (hunk 2 above)
Comment (ravinderjangra) [reply src/data_handler/idata_handler.rs:39]: +1

File: src/data_handler/idata_handler.rs (hunk 3 above)
Comment (ravinderjangra) [reply src/data_handler/idata_handler.rs:216]: `request_chunk_from_holders`

File: src/data_handler/idata_handler.rs (hunk 4 above)
Comment (ravinderjangra) [reply src/data_handler/idata_handler.rs:306]: Will rename to -> `list_chunks_for_duplication`

File: src/data_handler/idata_handler.rs (hunk 5 above)
Comment (ravinderjangra) [reply src/data_handler/idata_handler.rs:403]: Remove the block and return `None` and add a debug log.

File: src/data_handler/idata_handler.rs (hunk 6 above)
Comment (ravinderjangra) [reply src/data_handler/idata_handler.rs:407]: Will remove this.

File: src/data_handler/idata_handler.rs (hunk 8 above)
Comment (ravinderjangra) [reply src/data_handler/idata_handler.rs:549]: will rename to -> `get_metadata_for_chunks_held_by`

File: src/data_handler/idata_handler.rs (hunk 9 above)
Comment (ravinderjangra) [reply src/data_handler/idata_handler.rs:559]: updated msg to -> `not a valid chunkmetadata value`

File: src/data_handler/idata_handler.rs (hunk 11 above)
Comment (ravinderjangra) [reply src/data_handler/idata_handler.rs:650]: rename to -> `get_holders_for_chunk`

File: src/data_handler/idata_handler.rs (hunk 12 above)
Comment (ravinderjangra) [reply src/data_handler/idata_handler.rs:679]: rename to -> `get_new_holders_for_chunk`

File: src/data_handler.rs [hunk 13]
```diff
@@ -102,6 +102,12 @@ impl DataHandler {
             IData(idata_req) => {
                 match idata_req {
                     IDataRequest::Put(data) => {
+                        match requester {
+                            PublicId::Node(_) => info!("Got a pub idata request from Node"),
+                            PublicId::Client(_) => info!("Got a pub idata request from Client"),
+                            PublicId::App(_) => info!("Got a pub idata request from App"),
+                        };
```
Comment (ravinderjangra) [inline src/data_handler.rs:109]: Will remove these comments

File: src/data_handler.rs (hunk 1 above)
Comment (ravinderjangra) [reply src/data_handler.rs:246]: rename to -> `trigger_chunk_duplication`, add the comments explaning the functionality.

Update the info! -> trace!

File: src/vault.rs [hunk 14]
```diff
@@ -351,6 +351,18 @@ impl<R: CryptoRng + Rng> Vault<R> {
                 );
                 self.handle_routing_message(src, content)
             }
+            RoutingEvent::MemberLeft { name, age: _u8 } => {
+                info!("A node has left the section. Node: {:?}", name);
```
Comment (ravinderjangra) [inline src/vault.rs:355]: update to use `trace!`

**Pull request merged**